src/celeritas/neutron/model/detail/NuclearZoneBuilder.hh

```diff
@@ -0,0 +1,408 @@
+
+    CollectionBuilder<ZoneComponent> components_;
+    CollectionBuilder<NuclearZones, MemSpace::host, IsotopeId> zones_;
+
+    //// COMMON PROPERTIES ////
+
+    static CELER_CONSTEXPR_FUNCTION real_type half() { return 0.5; }
+    static CELER_CONSTEXPR_FUNCTION real_type four_thirds_pi()
+    {
+        return 4 * constants::pi / real_type{3};
+    }
+};
+
+//---------------------------------------------------------------------------//
+// INLINE DEFINITIONS
+//---------------------------------------------------------------------------//
+/*!
+ * Construct with cascade options and data.
+ */
+CELER_FUNCTION
+NuclearZoneBuilder::NuclearZoneBuilder(CascadeOptions const& options,
+                                       NeutronInelasticScalars const& scalars,
+                                       Data* data)
+    : options_(options)
+    , skin_depth_(0.611207 * options.radius_scale)
+    , neutron_mass_(scalars.neutron_mass)
+    , proton_mass_(scalars.proton_mass)
+    , components_(&data->components)
+    , zones_(&data->zones)
+{
+    CELER_EXPECT(options_);
+    CELER_EXPECT(data);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Construct nuclear zone data for a single isotope.
+ */
+void NuclearZoneBuilder::operator()(IsotopeView const& target)
+{
+    AtomicNumber a = target.atomic_mass_number();
+    size_type num_zones = this->num_nuclear_zones(a);
+
+    std::vector<ZoneComponent> comp(num_zones);
+    this->calc_zone_component(target, comp);
+
+    NuclearZones nucl_zones;
+    nucl_zones.num_zones = num_zones;
+    nucl_zones.zones = components_.insert_back(comp.begin(), comp.end());
+    zones_.push_back(nucl_zones);
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Return the number of nuclear zones.
+ */
+size_type NuclearZoneBuilder::num_nuclear_zones(AtomicMassNumber a)
+{
+    if (a < AtomicMassNumber{5})
+    {
+        return 1;
+    }
+    if (a < AtomicMassNumber{100})
+    {
+        return 3;
+    }
+    return 6;
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Calculate components of nuclear zone data: the nuclear zone radius, volume,
+ * density, Fermi momentum and potential function as in G4NucleiModel and as
+ * documented in section 24.2.3 of the Geant4 Physics Reference (release 11.2).
+ */
+CELER_FUNCTION
+void NuclearZoneBuilder::calc_zone_component(IsotopeView const& target,
+                                             ComponentVec& components)
+{
+    AtomicNumber a = target.atomic_mass_number();
+
+    // Calculate nuclear radius
+    real_type nuclear_radius = this->calc_nuclear_radius(a);
+
+    real_type skin_ratio = nuclear_radius / skin_depth_;
+    real_type skin_decay = std::exp(-skin_ratio);
+
+    // Temporary data for the zone-by-zone density function
+    size_type num_of_zones = components.size();
+    std::vector<real_type> integral(num_of_zones);
```

Comment:
Instead of doing the "number of zones"  calculation in a separate function and expecting the size to match in the `if` statements below, I think it'd be better to resize the vector inside this function. Inside the `if` statements, resize the vector to 1, `y.size()`, `alpha.size()`, etc., and then loop over `integral.size()` rather than a separate `num_of_zones` values.

Reply:
Not sure about this.  The number of zones is used for other properties (i.e., volumes, etc), so it may be still more readable keeping it as a common size across the whole code, which is called just once for a specific Isotope.

Reply:
Sorry, i didn't see this wasn't addressed. Are you expecting `NuclearZoneBuilder::num_nuclear_zones` to be called outside of that helper class? The number of zones is implicitly available through the components size, so the separate assumptions between `num_nuclear_zones` and `calc_zone_component` is a dangerous redundancy.